source/fuzz/transformation_swap_two_functions.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2021 Shiyu Liu 
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_swap_two_functions.h"
+
+#include "source/opt/function.h"
+#include "source/opt/module.h"
+#include <iostream>
+#include <string>
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationSwapTwoFunctions::TransformationSwapTwoFunctions(
+    protobufs::TransformationSwapTwoFunctions message)
+    : message_(std::move(message)) {}
+
+TransformationSwapTwoFunctions::TransformationSwapTwoFunctions(uint32_t id1, uint32_t id2) {
+  message_.set_function_id1(id1);
+  message_.set_function_id2(id2);
+}
+
+bool TransformationSwapTwoFunctions::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+
+  // Debugging
+  // for(auto& func: *ir_context->module()) {
+  //   std::cout<<func.result_id()<<" "<<func.PrettyPrint()<<std::endl;
+  // }
+  assert(message_.function_id1()!=message_.function_id2() && "Two functions cannot be the same.");
+  assert( (ir_context->GetFunction(message_.function_id1())!=nullptr 
+     || ir_context->GetFunction(message_.function_id2())!=nullptr) && "Both functions are not in range.");
+  assert(ir_context->GetFunction(message_.function_id1())!=nullptr && "Function 1 is not in range.");
+  assert(ir_context->GetFunction(message_.function_id2())!=nullptr && "Function 2 is not in range."); 
+
+
+  return true;
+}
+
+void TransformationSwapTwoFunctions::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // Found the two functions in ir_context and swap their position. 
+
+  opt::Function* func1_ptr = ir_context->GetFunction(message_.function_id1());
+  opt::Function* func2_ptr = ir_context->GetFunction(message_.function_id2());
+
+  assert( func1_ptr!=nullptr  && "ERROR: Function 1 was not found with the given id."); 
+  assert( func2_ptr!=nullptr  && "ERROR: Function 2 was not found with the given id.");
+  assert( &func1_ptr != &func2_ptr && "ERROR: Two functions cannot be the same.");
+
+  std::swap(func1_ptr, func2_ptr); 
```

Comment:
This does not quite work; it just swaps the local variables.

Even though we suggested using GetFunction, it turns out that swapping two functions is more easily done using iterators. So please could you obtain two iterators (e.g. `function_1` and `function_2`), and then do:

```
std::iter_swap(function_1.Get(), function_2.Get());
```

Or:

```
std::swap(*function_1.Get(), *function_2.Get());
```


The `.Get()` method returns the underlying iterator from the `UptrVectorIterator`, which is an iterator over `vector<unique_ptr<...>>`. And the `unique_ptr` type supports `swap`.

Reply:
Fixed. Now function iterator is attained by calling `GetFunctionIterator(id)` from `fuzzerutil` and use `std::iter_swap` to perform function swaps.